src/System.Management.Automation/engine/CommandCompletion/CompletionAnalysis.cs

```diff
@@ -94,26 +95,16 @@ private static bool IsCursorOutsideOfExtent(IScriptPosition cursor, IScriptExten
             return cursor.Offset < extent.StartOffset || cursor.Offset > extent.EndOffset;
         }
 
-        internal CompletionContext CreateCompletionContext(PowerShell powerShell)
+        internal static (Token TokenAtCursor, Token TokenBeforeCursor, List<Ast> RelatedAsts, int ReplacementIndex) ExtractAstContext(Ast inputAst, Token[] inputTokens, IScriptPosition cursor)
```

Comment:
I intentionally chose value tuple instead of a class to reduce short-living objects when `PredictInput` is called.
As for the field names, I will change it to camelCase.

Reply:
Actually, I take that back. It can be a struct :) I will update the code.

Reply:
Yeah a struct makes sense to me!